### PR TITLE
[WIP][feature] googledrive

### PR DESCRIPTION
**Pull request by icereval**

### CODE REVIEW

File: website/addons/box/model.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+
+    # Box user id, for example, "4974056"
+    user_id = fields.StringField(primary=True, required=True)
+    # Box user name this is the user's login
+    username = fields.StringField()
+    _access_token = fields.StringField()
+    refresh_token = fields.StringField()
+    expires_at = fields.DateTimeField()
+
+    @property
```
Comment (sloria) [inline website/addons/box/model.py]: Given that this method may make an HTTP request (and therefore raise an HTTP error) and that  properties have strange error handling behavior, we might want to make this a method (maybe `fetch_access_token`.

That way, we could also rename `_access_token` to `access_token` and get rid of the setter.

This does not necessarily need to happen before this is merged.

Comment (icereval) [inline website/addons/box/model.py]: fair request, i'll make the change

File: website/addons/box/routes.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        Rule(
+            '/settings/box/oauth/',
+            'get',
+            views.auth.box_oauth_start,  # Use same view func as node oauth start
+            json_renderer,
+            endpoint_suffix='_user'          # but add a suffix for url_for
+        ),
+
+        Rule(
+            '/addons/box/oauth/finish/',
```
Comment (sloria) [inline website/addons/box/routes.py]: For consistency with @lyndsysimon 's Oauth system which we will likely migrate to in the near future, let's use `/oauth/callback/box/`.

Comment (sloria) [inline website/addons/box/routes.py]: The same change should also be made to googledrive

File: website/addons/box/static/boxFangornConfig.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (sloria) [inline website/addons/box/static/boxFangornConfig.js]: Add `'use strict'`

File: website/addons/box/static/boxNodeConfig.js [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    };
+
+    /**
+        * Send DELETE request to deauthorize this node.
+        */
+    function sendDeauth() {
+        return $.ajax({
+            url: self.urls().deauthorize,
+            type: 'DELETE',
+            success: function() {
```
Comment (sloria) [inline website/addons/box/static/boxNodeConfig.js]: Use the Promise interface.

File: website/addons/box/static/boxNodeConfig.js [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+                // Update observables
+                self.nodeHasAuth(false);
+                self.cancelSelection();
+                self.currentDisplay(null);
+                self.changeMessage('Deauthorized Box.', 'text-warning', 3000);
+            },
+            error: function() {
+                self.changeMessage('Could not deauthorize because of an error. Please try again later.',
```
Comment (sloria) [inline website/addons/box/static/boxNodeConfig.js]: Log this failure to sentry with Raven.

File: website/addons/box/static/boxUserConfig.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    self.validCredentials = ko.observable(true);
+    self.boxName = ko.observable();
+    self.urls = ko.observable({});
+    // Whether the initial data has been loaded.
+    self.loaded = ko.observable(false);
+    self.nNodesAuthorized = 0;
+    // Update above observables with data from server
+    $.ajax({
+        url: url, type: 'GET', dataType: 'json',
+        success: function(response) {
```
Comment (sloria) [inline website/addons/box/static/boxUserConfig.js]: Use the `Promise` interface.

File: website/addons/box/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+            'token_type': 'something',
+            'access_token': 'something',
+            'refresh_token': 'something'
+        }
+        mock_oauth.return_value = ('mytoken123', 'myboxid', 'done')
+        url = api_url_for('box_oauth_finish')
+
+        res = self.app.get(url)
+
+        assert_is_redirect(res)
```
Comment (sloria) [inline website/addons/box/tests/test_views.py]: This test could be made more complete, e.g. testing that the redirect URL is correct, the addon was added to the user, etc. @mfraezz Can you please work on this post-merge?

File: website/addons/box/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+        mock_account_info.return_value = {'display_name': 'Mr. Box'}
+        url = api_url_for('box_user_config_get')
+        res = self.app.get(url, auth=self.user.auth)
+        assert_equal(res.status_code, 200)
+        # The JSON result
+        result = res.json['result']
+        assert_true(result['userHasAuth'])
+
+    @mock.patch('website.addons.box.client.BoxClient.get_user_info')
+    @mock.patch('website.addons.box.model.BoxNodeSettings.full_folder_path')
```
Comment (sloria) [inline website/addons/box/tests/test_views.py]: @mfraezz Instead of mocking `full_folder_path`, can you set `_folder_data` in this test?

File: website/addons/box/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+        mock_account_info.return_value = {'display_name': 'Mr. Box'}
+        url = api_url_for('box_user_config_get')
+        res = self.app.get(url, auth=self.user.auth)
+        assert_equal(res.status_code, 200)
+        # The JSON result
+        result = res.json['result']
+        assert_true(result['validCredentials'])
+
+    @mock.patch('website.addons.box.client.BoxClient.get_user_info')
+    @mock.patch('website.addons.box.model.BoxNodeSettings._fetch_folder_data')
```
Comment (sloria) [inline website/addons/box/tests/test_views.py]: @mfraezz For this test and subsequent tests, could you set `_folder_data` instead of mocking `_fetch_folder_data`?

File: website/addons/box/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+
+    def setUp(self):
+        super(TestFilebrowserViews, self).setUp()
+        self.user.add_addon('box')
+        settings = self.user.get_addon('box')
+        oauth = BoxOAuthSettings(user_id='not none', _access_token='Nah')
+        oauth.save()
+        settings.oauth_settings = oauth
+        settings.save()
+        self.patcher = mock.patch('website.addons.box.model.BoxNodeSettings.folder')
```
Comment (sloria) [inline website/addons/box/tests/test_views.py]: @mfraezz Related to previous comments, is this patcher necessary if `_folder_data` is set?

File: website/addons/box/utils.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        self.node.add_log(
+            action="box_{0}".format(action),
+            params=params,
+            auth=self.auth
+        )
+        if save:
+            self.node.save()
+
+
+def handle_box_error(error, msg):
```
Comment (sloria) [inline website/addons/box/utils.py]: `msg` is unused here.

Comment (sloria) [inline website/addons/box/utils.py]: Will fix during merge.

File: website/addons/box/settings/__init__.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import logging
+from .defaults import *  # noqa
+
+try:
+    from .local import *  # noqa
+except ImportError as error:
+    logging.warn('No local.py settings file found')
```
Comment (sloria) [inline website/addons/box/settings/__init__.py]: Use a named logger instead of the root logger.

```python
import logging
from .defaults import *  # noqa

logger = logging.getLogger(__name__)

try:
    from .local import *  # noqa
except ImportError as error:
    logger.warn('No local.py settings file found')
```

Will make this change during merge.

File: website/addons/box/templates/box_node_settings.mako [hunk 13]
```diff
@@ -0,0 +1,138 @@
+
+<link rel="stylesheet" href="/static/addons/box/box.css">
```
Comment (sloria) [inline website/addons/box/templates/box_node_settings.mako]: With the addition of `addonsettings.css`, these addon-specific css files should no longer be necessary. `btn-box` should be replaced with `btn-addon`. 

@mfraezz Can you make this change post-merge?

File: website/addons/box/templates/box_view_file.mako [hunk 14]
```diff
@@ -0,0 +1,82 @@
+<%inherit file="project/addon/view_file.mako" />
+
+<%def name="file_versions()">
```
Comment (sloria) [inline website/addons/box/templates/box_view_file.mako]: @chrisseto Is this file necessary given your changes to file detail pages?

File: website/addons/box/tests/test_models.py [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+    def test_provider(self):
+        assert_equal('box', BoxFile().provider)
+
+    def test_correct_path(self):
+        guid = BoxFile(node=self.project, path='1234567890/foo/bar')
+
+        assert_equals(guid.path, '1234567890/foo/bar')
+        assert_equals(guid.waterbutler_path, '/1234567890/foo/bar')
+
+    @mock.patch('website.addons.base.requests.get')
```
Comment (sloria) [inline website/addons/box/tests/test_models.py]: @mfraezz Why is `requests.get` mocked here?

File: website/addons/box/tests/test_views.py (hunk 9 above)
Comment (chrisseto) [inline website/addons/box/tests/test_views.py]: I made this change, Its easier to mock `_fetch_folder_data` because the return value can just be set and flake8 likes it a bit more.

File: website/addons/box/static/boxNodeConfig.js (hunk 4 above)
Comment (sloria) [inline website/addons/box/static/boxNodeConfig.js]: This change will be made during merge.

File: website/addons/box/static/boxFangornConfig.js (hunk 3 above)
Comment (sloria) [inline website/addons/box/static/boxFangornConfig.js]: Made during merge.

File: website/addons/box/tests/test_views.py (hunk 8 above)
Comment (chrisseto) [inline website/addons/box/tests/test_views.py]: I made this change as well, the request that `_fetch_folder_data` makes just needs to be stopped. Mocking `_folder_data` it just another step down.

File: website/addons/box/templates/box_view_file.mako (hunk 14 above)
Comment (chrisseto) [inline website/addons/box/templates/box_view_file.mako]: No longer needed, I'll remove it.

File: website/addons/box/routes.py (hunk 2 above)
Comment (icereval) [inline website/addons/box/routes.py]: done
